source/vstd/std_specs/hash.rs

```diff
@@ -428,6 +428,234 @@ pub fn ex_hash_map_clear<Key, Value, S>(m: &mut HashMap<Key, Value, S>)
     m.clear()
 }
 
+// We now specify the behavior of `HashSet`.
+#[verifier::external_type_specification]
+#[verifier::external_body]
+#[verifier::reject_recursive_types(Key)]
+#[verifier::reject_recursive_types(S)]
+pub struct ExHashSet<Key, S>(HashSet<Key, S>);
+
+impl<Key, S> View for HashSet<Key, S> {
+    type V = Set<Key>;
+
+    #[verifier::external_body]
+    closed spec fn view(&self) -> Set<Key>;
+}
+
+pub open spec fn spec_hash_set_len<Key, S>(m: &HashSet<Key, S>) -> usize;
+
+pub broadcast proof fn axiom_spec_hash_set_len<Key, S>(m: &HashSet<Key, S>)
+    ensures
+        obeys_key_model::<Key>() && builds_valid_hashers::<S>() ==> #[trigger] spec_hash_set_len(m)
+            == m@.len(),
+{
+    admit();
+}
+
+#[verifier::external_fn_specification]
+#[verifier::when_used_as_spec(spec_hash_set_len)]
+pub fn ex_hash_set_len<Key, S>(m: &HashSet<Key, S>) -> (len: usize)
+    ensures
+        len == spec_hash_set_len(m),
+{
+    m.len()
+}
+
+#[verifier::external_fn_specification]
+pub fn ex_hash_set_new<Key>() -> (m: HashSet<Key, RandomState>)
+    ensures
+        m@ == Set::<Key>::empty(),
+{
+    HashSet::<Key>::new()
+}
+
+#[verifier::external_fn_specification]
+pub fn ex_hash_set_with_capacity<Key>(capacity: usize) -> (m: HashSet<Key, RandomState>)
+    ensures
+        m@ == Set::<Key>::empty(),
+{
+    HashSet::<Key>::with_capacity(capacity)
+}
+
+#[verifier::external_fn_specification]
+pub fn ex_hash_set_reserve<Key, S>(m: &mut HashSet<Key, S>, additional: usize) where
+    Key: Eq + Hash,
+    S: BuildHasher,
+
+    ensures
+        m@ == old(m)@,
+{
+    m.reserve(additional)
+}
+
+#[verifier::external_fn_specification]
+pub fn ex_hash_set_insert<Key, S>(m: &mut HashSet<Key, S>, k: Key) -> (result: bool) where
+    Key: Eq + Hash,
+    S: BuildHasher,
+
+    ensures
+        obeys_key_model::<Key>() && builds_valid_hashers::<S>() ==> {
+            &&& m@ == old(m)@.insert(k)
+            &&& match result {
+                true => old(m)@.contains(k),
```

Comment:
This is backwards. `HashSet::insert` returns a boolean indicating if the element *wasn't* in the set before.

Reply:
```suggestion
        result == ! old(m)@.contains(k)
```

Reply:
```suggestion
result == ! old(m)@.contains(k)
```